packages/common/src/api/tan-query/usePurchases.ts

```diff
@@ -0,0 +1,104 @@
+import { useMemo } from 'react'
+
+import { full } from '@audius/sdk'
+import { useInfiniteQuery } from '@tanstack/react-query'
+
+import { Id } from '~/api/utils'
+import { useAudiusQueryContext } from '~/audius-query'
+import { ID } from '~/models'
+import {
+  USDCContentPurchaseType,
+  USDCPurchaseDetails
+} from '~/models/USDCTransactions'
+import { Nullable } from '~/utils/typeUtils'
+
+import { QUERY_KEYS } from './queryKeys'
+import { useCollections } from './useCollections'
+import { useTracks } from './useTracks'
+import { useUsers } from './useUsers'
+import { parsePurchase } from './utils/parsePurchase'
+
+export type GetPurchaseListArgs = {
+  userId: Nullable<ID>
+  sortMethod?: full.GetPurchasesSortMethodEnum
+  sortDirection?: full.GetPurchasesSortDirectionEnum
+}
+
+export type PaginatedPurchaseListArgs = GetPurchaseListArgs & {
+  limit: number
+  offset: number
+}
+
+export const usePurchases = (
+  args: GetPurchaseListArgs,
+  options: { pageSize: number; enabled?: boolean }
+) => {
+  const { audiusSdk } = useAudiusQueryContext()
+  const { pageSize, enabled } = options
+  const queryResult = useInfiniteQuery({
+    queryKey: [QUERY_KEYS.purchases, args],
+    enabled: enabled !== false && !!args.userId,
+    initialPageParam: 0,
+    getNextPageParam: (
+      lastPage: USDCPurchaseDetails[],
+      allPages: USDCPurchaseDetails[][]
+    ) => {
+      if (lastPage.length < pageSize) return undefined
+      return allPages.length * pageSize
+    },
+    queryFn: async ({ pageParam }) => {
+      const sdk = await audiusSdk()
+      const { data = [] } = await sdk.full.users.getPurchases({
+        limit: pageSize,
+        offset: pageParam,
+        sortDirection: args.sortDirection,
+        sortMethod: args.sortMethod,
+        id: Id.parse(args.userId!),
+        userId: Id.parse(args.userId!)
+      })
+
+      return data.map(parsePurchase)
+    }
+  })
+
+  const { data, fetchNextPage, hasNextPage, isFetchingNextPage, ...rest } =
+    queryResult
+
+  // Flatten pages into a single array
+  const flatData = useMemo(() => data?.pages.flat() ?? [], [data?.pages])
+
+  // Pre-fetch purchased content metadata
+  const userIdsToFetch = useMemo(
+    () => flatData.map(({ buyerUserId }) => buyerUserId),
+    [flatData]
+  )
+  const trackIdsToFetch = useMemo(
+    () =>
+      flatData
+        .filter(
+          ({ contentType }) => contentType === USDCContentPurchaseType.TRACK
+        )
+        .map(({ contentId }) => contentId),
+    [flatData]
+  )
+  const collectionIdsToFetch = useMemo(
+    () =>
+      flatData
+        .filter(
+          ({ contentType }) => contentType === USDCContentPurchaseType.ALBUM
+        )
+        .map(({ contentId }) => contentId),
+    [flatData]
+  )
+  // Call the hooks dropping results to pre-fetch the data
+  useUsers(userIdsToFetch)
```

Comment:
fwiw I think the original way this hook worked is that it didn't return from the fetch function until the dependent entities were also fetched. I suspect it might be fine if the components inside the table are also using tan-query now. But just wanted to call that out.

Also curious why this pattern instead of using the specific prefetch functionality on tq?

Reply:
yes same question about the other hook!

Reply:
after some research, I think it's best for us to just fire and forget the hook here

If the multi-fetching race condition becomes an issue, we are probably not benefiting from the pre-fetching anyway and can either address the race or remove this.

The built-in prefetching requires us to pass the fetchFunction, and I think in this case since we're already in a hook, this is more ergonomic. The pre-fetching may be useful during write operations or other cases where we antcipate need before it arises. I just don't think it's necessary for parent child structures where we can just call the hooks normally.